File: examples/ruby/printMath.rb
#!/usr/bin/env ruby
#
# @file    printMath.rb
# @brief   Prints Rule, Reaction, and Event formulas in a given SBML Document
# @author  Alex Gutteridge (Ruby conversion of examples/c/printMath.c)
# @author  Ben Bornstein
#
# $Id$
# $Source$
#
# This file is part of libSBML.  Please visit http://sbml.org for more
# information about SBML, and the latest version of libSBML.
#

require 'libSBML'
     
module LibSBML
  class Model
    def printMath
      getNumFunctionDefinitions.times do |n|
        puts "Function #{n}: #{getFunctionDefinition(n)}" 
      end
      getNumRules.times do |n|
        puts "Rule #{n}: #{getRule(n)}"
      end  
      getNumReactions.times do |n|
        puts "Reaction #{n}: #{getReaction(n)}"
      end  
      getNumEvents.times do |n|
        puts "Event #{n} #{getEvent(n)}"
      end
    end
  end
  
  class FunctionDefinition
     def to_s
       s = ""
       if isSetMath
         s << "#{getId}("
         math = getMath
         
         #Print function args
         if math.getNumChildren > 1
           s << math.getLeftChild.getName
           ((math.getNumChildren)-2).times do |n|
             s << ", " + math.getChild(n+1).getName
           end
         end
         
         s << ") := "

         #Print function body
         if math.getNumChildren == 0
           s << "(no body defined)"
         else
           s << LibSBML::formulaToString(math.getChild(math.getNumChildren-1))
         end

         return s

       end
     end
  end
  class Rule
     def to_s
        if isSetMath
          LibSBML::formulaToString(getMath)
        end  
     end
  end
  class Reaction
    def to_s
      if isSetKineticLaw and getKineticLaw.isSetMath
         LibSBML::formulaToString(getKineticLaw.getMath)                                           
      end
    end
  end
  class EventAssignment
    def to_s
      if isSetMath
        "#{getVariable} = #{LibSBML::formulaToString(getMath)}" 
      end
    end
  end
  class Event
    def to_s
      d = ''
      t = ''
      a = []
      if isSetDelay
        d = "Delay: #{LibSBML::formulaToString(getDelay.getMath)} "
      end
      if isSetTrigger
        t = "Trigger: #{LibSBML::formulaToString(getTrigger.getMath)} "
      end
      getNumEventAssignments.times do |n|
         a << getEventAssignment(n).to_s 
      end
      d + t + a.join(", ")   
    end
  end
end
 
if ARGV.size != 1
  puts "Usage: printMath filename"
  exit(1)
end

d = LibSBML::readSBML(ARGV[0])
d.printErrors

d.getModel.printMath

File: examples/python/printMath.py
#!/usr/bin/env python
## 
## @file    printMath.py
## @brief   Prints Rule, Reaction, and Event formulas in a given SBML Document
## @author  Ben Bornstein
## @author  Sarah Keating
## 
## 
## This file is part of libSBML.  Please visit http://sbml.org for more
## information about SBML, and the latest version of libSBML.
## 


import sys
import os.path
from libsbml import *

def printFunctionDefinition(n, fd):
    if (fd.isSetMath()):
        print("FunctionDefinition " + str(n) + ", " + fd.getId());

        math = fd.getMath();

        # Print function arguments. 
        if (math.getNumChildren() > 1):
            print("(" + (math.getLeftChild()).getName());

            for n in range (1, math.getNumChildren()):
                print(", " + (math.getChild(n)).getName());

        print(") := ");

        # Print function body. 
        if (math.getNumChildren() == 0):
            print("(no body defined)");
        else:
            math = math.getChild(math.getNumChildren() - 1);
            formula = formulaToString(math);
            print(formula + "\n");


def printRuleMath(n, r):
    if (r.isSetMath()):
        formula = formulaToString(r.getMath());

        if (len(r.getVariable()) > 0):
            print("Rule " + str(n) + ", formula: "
                             + r.getVariable() + " = " + formula + "\n");
        else:
            print("Rule " + str(n) + ", formula: "
                             + formula + " = 0" + "\n");


def printReactionMath(n, r):
    if (r.isSetKineticLaw()):
        kl = r.getKineticLaw();
        if (kl.isSetMath()):
            formula = formulaToString(kl.getMath());
            print("Reaction " + str(n) + ", formula: " + formula + "\n");


def printEventAssignmentMath(n, ea):
    if (ea.isSetMath()):
        variable = ea.getVariable();
        formula = formulaToString(ea.getMath());
        print("  EventAssignment " + str(n)
                              + ", trigger: " + variable + " = " + formula + "\n");


def printEventMath(n, e):
    if (e.isSetDelay()):
        formula = formulaToString(e.getDelay().getMath());
        print("Event " + str(n) + " delay: " + formula + "\n");

    if (e.isSetTrigger()):
        formula = formulaToString(e.getTrigger().getMath());
        print("Event " + str(n) + " trigger: " + formula + "\n");

    for i in range(0,e.getNumEventAssignments()):
        printEventAssignmentMath(i + 1, e.getEventAssignment(i));

    print;

def printMath(m):
    for n in range(0,m.getNumFunctionDefinitions()):
        printFunctionDefinition(n + 1, m.getFunctionDefinition(n));

    for n in range(0,m.getNumRules()):
        printRuleMath(n + 1, m.getRule(n));

	print;

    for n in range(0, m.getNumReactions()):
        printReactionMath(n + 1, m.getReaction(n));

    print;

    for n in range(0,m.getNumEvents()):
        printEventMath(n + 1, m.getEvent(n));


def main (args):
  """Usage: printMath filename
  """
  if (len(args) != 2):
      print("\n" + "Usage: printMath filename" + "\n" + "\n");
      return 1;
  
  filename = args[1];
  document = readSBML(filename);
  
  if (document.getNumErrors() > 0):
      print("Encountered the following SBML errors:" + "\n");
      document.printErrors();
      return 1;
  
  model = document.getModel();
  
  if (model == None):
      print("No model present." + "\n");
      return 1;
  
  printMath(model);
  print;
  return 0;

if __name__ == '__main__':
  main(sys.argv)  
